File: projects/11-vod-streaming/src/manifest.rs
//! V3 — Manifest generation: HLS `.m3u8` + DASH `.mpd`.
//!
//! The manifest is the index a player reads *before* any media: it names the init
//! segment, lists every media segment and its exact duration, and (at the master /
//! MPD level) advertises the bitrate ladder so the player can choose and switch.
//! HLS is a line-oriented text format (`#EXT-X-…` tags); DASH is XML with a
//! `SegmentTemplate`/`SegmentTimeline` model. Both describe the *same* segments V2
//! produced — this module is the mapping.
//!
//! These are pure functions (segment list → string): no I/O, trivially testable
//! against golden files.

use crate::isobmff::Track;
use crate::segment::SegmentIndex;

/// One rung of the ABR ladder, for the HLS master playlist / DASH adaptation set.
#[derive(Debug, Clone)]
pub struct RenditionInfo {
    /// Rendition id (e.g. `720p`) — also the path segment in its media-playlist URI.
    pub id: String,
    /// Peak bitrate in bits/sec — the HLS `BANDWIDTH` attribute. A player picks a
    /// rung from this vs. its measured throughput.
    pub bandwidth: u32,
    /// Coded resolution for the `RESOLUTION` attribute (0×0 ⇒ omit).
    pub width: u16,
    pub height: u16,
    /// URI of this rendition's media playlist, relative to the master.
    pub uri: String,
}

/// Render the **HLS media playlist** for one rendition (V3 core).
///
/// TODO(V3): emit, in order:
///   `#EXTM3U`
///   `#EXT-X-VERSION:7`                     (7+ is required for fMP4 / `EXT-X-MAP`)
///   `#EXT-X-TARGETDURATION:<ceil(max seg seconds)>`
///   `#EXT-X-MEDIA-SEQUENCE:0`
///   `#EXT-X-PLAYLIST-TYPE:VOD`
///   `#EXT-X-MAP:URI="init.mp4"`           (the init segment)
///   then per segment:
///     `#EXTINF:<seconds with enough precision>,`
///     `seg/<index>`                        (matches the delivery route)
///   `#EXT-X-ENDLIST`                       (VOD: the list is complete)
/// The summed `#EXTINF`s must equal the track duration within a frame — carry the
/// real per-segment durations from `index`, don't re-round the target.
pub fn hls_media_playlist(index: &SegmentIndex, track: &Track, _target_secs: f64) -> String {
    let _ = (index, track);
    todo!("V3: render the HLS media playlist (EXT-X-MAP, EXTINF per segment, ENDLIST)")
}

/// Render the **HLS master playlist** advertising the rendition ladder (V3/V4).
///
/// TODO(V3): emit `#EXTM3U` + `#EXT-X-VERSION:7`, then one
/// `#EXT-X-STREAM-INF:BANDWIDTH=<bps>,RESOLUTION=<w>x<h>` line per rendition
/// followed by its media-playlist `uri`. This is what makes ABR possible: the
/// player reads the ladder here and switches rungs as bandwidth changes.
pub fn hls_master_playlist(renditions: &[RenditionInfo]) -> String {
    let _ = renditions;
    todo!("V3: render the HLS master playlist (one EXT-X-STREAM-INF per rendition)")
}

/// Render the **DASH MPD** for one rendition's segments (V3).
///
/// TODO(V3): emit a `static` (VOD) MPD: `MPD` → `Period` → `AdaptationSet` →
/// `Representation` with the codec/bandwidth, and a `SegmentList` (or
/// `SegmentTemplate` + `SegmentTimeline`) referencing `init.mp4` and each
/// `seg/<index>` with its duration in the MPD timescale. Same segments as the HLS
/// playlist, described the DASH way — proving you understand both models map onto
/// the one segment list.
pub fn dash_mpd(index: &SegmentIndex, track: &Track) -> String {
    let _ = (index, track);
    todo!("V3: render the DASH MPD (Representation + SegmentList over the same segments)")
}

#[cfg(test)]
mod tests {
    // TODO(V3): golden-file tests against the committed fixture:
    //   - `hls_media_playlist` matches expected output byte-for-byte, and its
    //     summed EXTINF == track duration within one frame (`renders_hls_media_playlist`);
    //   - `hls_master_playlist` lists every rendition with BANDWIDTH + RESOLUTION;
    //   - `dash_mpd` validates against a DASH conformance checker (`renders_dash_mpd`).
}

File: projects/11-vod-streaming/src/vod_streaming/manifest.py
"""V3 — Manifest generation: HLS `.m3u8` + DASH `.mpd`.

Module: `src/vod_streaming/manifest.py`.

The manifest is the index a player reads *before* any media: it names the init
segment, lists every media segment with its exact duration, and — at the master /
MPD level — advertises the bitrate ladder so the player can choose a rung and
switch between them. HLS is a line-oriented text format of `#EXT-X-...` tags;
DASH is XML with a `SegmentTemplate`/`SegmentTimeline` model. Both describe the
*same* segments V2 produced, which makes this module a mapping exercise and the
cheapest place in the project to see that the two formats are the same idea
wearing different clothes.

These are pure functions — segment list in, string out. No I/O, no state, no
async: exactly the shape golden-file tests want, which is why the SPEC's Proof
for this vertical is a byte-for-byte comparison against committed expected output.

## Getting the text right in Python

**Build a list of lines and `"\\n".join(...)` it.** Repeated `+=` on a `str` is
the same quadratic trap as with `bytes`, and a list makes the tag order — which
is load-bearing in HLS — something you can read straight down the function.

**End with a newline.** `#EXT-X-ENDLIST` without a trailing `\\n` is accepted by
most players and rejected by Apple's `mediastreamvalidator`, which is the tool the
SPEC's Proof names. Cheap to get right, annoying to debug.

**Format `#EXTINF` from the tick count, not from accumulated seconds.**
`f"{entry.seconds(timescale):.6f}"` on each segment independently keeps the error
per-segment; summing floats as you go lets it compound, and the SPEC grades the
sum against the track duration within one frame. Three decimals is the usual
choice and is enough for a 90 kHz timescale; pick a precision and note it.

**Do not hand-write the XML.** `xml.etree.ElementTree` is in the standard library
and escapes attribute values for you. An f-string MPD works perfectly until an
asset is named `tom & jerry`, at which point it emits `&` raw and every DASH
parser rejects the document — a bug that looks like a manifest problem and is
actually a quoting problem. `ElementTree.tostring(root, encoding="unicode")` also
gives you well-formedness for free, which is half of what the conformance
validator checks.

**DASH durations are ISO 8601, not seconds.** `mediaPresentationDuration` wants
`PT1M30.5S`, not `90.5`. There is no formatter for this in the standard library —
`datetime.timedelta` does not produce it — so write the handful of lines and unit
test them, including the zero case and durations over an hour.
"""

from __future__ import annotations

from dataclasses import dataclass

from .isobmff import Track
from .segment import SegmentIndex

__all__ = ["RenditionInfo", "dash_mpd", "hls_master_playlist", "hls_media_playlist"]


@dataclass(slots=True)
class RenditionInfo:
    """One rung of the ABR ladder, for the HLS master playlist / DASH adaptation set."""

    id: str
    """Rendition id (e.g. `720p`) — also the path segment in its media-playlist URI."""

    bandwidth: int
    """Peak bitrate in bits/sec — the HLS `BANDWIDTH` attribute.

    A player compares this against its measured throughput to pick a rung, so it
    should be the *peak*, not the average: a rendition that averages 3 Mbps but
    spikes to 6 will stall a client that budgeted for 3."""

    width: int = 0
    """Coded resolution for the `RESOLUTION` attribute. 0x0 means omit the
    attribute entirely rather than emitting `RESOLUTION=0x0`, which is invalid."""

    height: int = 0

    uri: str = ""
    """URI of this rendition's media playlist, relative to the master."""


def hls_media_playlist(index: SegmentIndex, track: Track) -> str:
    """Render the **HLS media playlist** for one rendition — the V3 core.

    TODO(V3): emit, in this order:

        #EXTM3U
        #EXT-X-VERSION:7                    (7+ is required for fMP4 / EXT-X-MAP)
        #EXT-X-TARGETDURATION:<ceil(longest segment, seconds)>
        #EXT-X-MEDIA-SEQUENCE:0
        #EXT-X-PLAYLIST-TYPE:VOD
        #EXT-X-MAP:URI="init.mp4"           (the init segment)

    then, per segment:

        #EXTINF:<seconds>,
        seg/<index>                         (matches the delivery route)

    and finally `#EXT-X-ENDLIST`, which is what tells a player this is VOD and
    complete rather than a live playlist it should keep re-fetching.

    The `#EXTINF` line ends with a comma — the tag's grammar is
    `#EXTINF:<duration>,<title>` and the title is optional but the comma is not.
    Omitting it is the classic "why does nothing play" playlist bug.

    `#EXT-X-TARGETDURATION` comes from `index.target_duration(track.timescale)`,
    which already rounds up correctly. The summed `#EXTINF`s must equal the track
    duration within one frame — carry the real per-segment durations, never
    `target_secs` repeated.
    """
    del index, track
    raise NotImplementedError("V3: render the HLS media playlist (EXT-X-MAP, EXTINF, ENDLIST)")


def hls_master_playlist(renditions: list[RenditionInfo]) -> str:
    """Render the **HLS master playlist** advertising the rendition ladder (V3/V4).

    TODO(V3): emit `#EXTM3U` and `#EXT-X-VERSION:7`, then per rendition a
    `#EXT-X-STREAM-INF:BANDWIDTH=<bps>,RESOLUTION=<w>x<h>` line immediately
    followed by its media-playlist `uri` on the next line. The pairing is
    positional — the URI belongs to the tag above it — which is why this is a
    playlist and not a list of attributes.

    This file is what makes ABR possible: the player reads the ladder here and
    switches rungs as its bandwidth estimate moves. Order the rungs deliberately
    (a player commonly starts with the first) and say why in `docs/11-design.md`.

    Consider what to do with a rendition whose `bandwidth` is still 0 because
    nothing has probed it yet — an advertised `BANDWIDTH=0` tells a player this
    rung is free, and it will pick it every time.
    """
    del renditions
    raise NotImplementedError("V3: render the HLS master playlist (one EXT-X-STREAM-INF each)")


def dash_mpd(index: SegmentIndex, track: Track) -> str:
    """Render the **DASH MPD** for one rendition's segments (V3).

    TODO(V3): emit a `static` (VOD) MPD — `MPD` -> `Period` -> `AdaptationSet` ->
    `Representation` with the codec and bandwidth, and a `SegmentList` (or
    `SegmentTemplate` + `SegmentTimeline`) referencing `init.mp4` and each
    `seg/<index>` with its duration. The same segments as the HLS playlist,
    described the DASH way — building both is what proves the two models map onto
    one segment list.

    Details that a conformance validator will fail you on:

      * the `xmlns` (`urn:mpeg:dash:schema:mpd:2011`) and `profiles`
        (`urn:mpeg:dash:profile:isoff-on-demand:2011` for VOD) attributes;
      * `mediaPresentationDuration` in ISO 8601 — see the module docstring;
      * `@timescale` on the segment element, so durations stay integers in the
        track's own ticks rather than being rounded into seconds;
      * the `codecs` string (`avc1.64001f`), which is assembled from the SPS
        bytes V1 kept in `Track.codec.setup` — profile, constraint flags and level
        as six hex digits. HLS wants the same string in `CODECS` on
        `#EXT-X-STREAM-INF`, so build it once somewhere both can reach.

    Use `xml.etree.ElementTree`; see the module docstring on why not an f-string.
    """
    del index, track
    raise NotImplementedError("V3: render the DASH MPD over the same segment list")
